docs/design.md

```diff
@@ -0,0 +1,54 @@
+## Tableroll Design
+
+### File descriptor handoff protocol
+
+tableroll defines a custom protocol in order to coordinate taking ownership of a collection of file descriptors over a unix socket. It consists of the following notable filesystem paths:
+
+* The coordination directory &mdash; this directory, typically in
+  `/run/${programName}/tableroll`, is used to coordinate state between each
+  running process involved in the tableroll upgrade group.
+* The pid file &mdash; a `pid` file within the coordination directory is used
+  to indicate the current active holder of all file descriptors. A process will
+  write its own pid to that file after a handoff with the previous process
+  whose pid was in the pid file.
+* Coordination unix sockets &mdash; each process will also listen on a socket
+  within the coordination directory. This socket is the means by which a file
+  descriptor handoff may be initiated, and the medium over which file
+  descriptors will be passed. Each socket is named `${pid}.sock` within the
+  coordination directory.
+
+#### Handoff protocol
+
+The actual protocol always takes place between two processes. For the sake of
+example, let's assume the chosen coordination directory is
+`/run/example/tableroll`, and that two processes using that directory exist.
+These processes will be the "first" and the "second" processes, where the
+"first" process arbitrarily holds two open file descriptors that are marked
+passable, '127.0.0.1:80' and '127.0.0.1:443'.
+
+The handoff of these two FDs from "first" to "second" will look like the following:
+
+1. "second" begins listening on `/run/example/tableroll/${second_pid}.sock`.
+1. "second" takes an exclusive lock on the pid file, `/run/example/tableroll/pid`.
+1. "second" reads the value `${first_pid}` from the pid file.
+1. "second" opens a unix connection to `/run/example/tableroll/${first_pid}.sock`.
+1. "first" accepts the connection and writes the following data to it:
+    1. `70` &mdash; The length of the following message, as a 4 byte signed integer in big endian.
+    1. `[["listener","tcp","127.0.0.1:80"],["listener","tcp","127.0.0.1:443"]]` &mdash; A JSON encoding of the names of all file descriptors to expect
```

Comment:
Inherited from them, I would have personally used a struct with 3 string members. [xref](https://github.com/cloudflare/tableflip/blob/d1e14f24db7d216f82bde697a0c6551de4125e2d/fds.go#L25-L35).

It's a fixed size array so it can be a key in a map. The choice of `struct { kind string; network tring; addr string }` or `[3]string` seems pretty arbitrary to me.